internal/xds/translator/translator.go

```diff
@@ -125,65 +126,150 @@ func (t *Translator) processHTTPListenerXdsTranslation(tCtx *types.ResourceVersi
 			}
 		}
 
-		// Allocate virtual host for this httpListener.
-		// 1:1 between IR HTTPListener and xDS VirtualHost
-		vHost := &routev3.VirtualHost{
-			Name:    httpListener.Name,
-			Domains: httpListener.Hostnames,
-		}
 		protocol := DefaultProtocol
 		if httpListener.IsHTTP2 {
 			protocol = HTTP2
 		}
 
-		// Check if an extension is loaded that wants to modify xDS Routes after they have been generated
+		routesByHostname := make(map[string][]*ir.HTTPRoute, len(httpListener.Routes))
```

Comment:
trying to understand why we need this map (sidenote - maps lack order, making any sort of comparison really hard)
instead why not just
```
         for _, httpRoute := range httpListener.Routes {
 		// 1:1 between IR HTTPRoute and xDS VirtualHost
 		vHost := &routev3.VirtualHost{
 			Name:    httpRoute. Name,
 			Domains: httpRoute.Hostnames, // this should be an intersection of listener and route hostnames from the gateway api layer
 		}
```

Reply:
a virtual host needs to contain all routes that match into it eg if the vhost domain is *.google.com we need routes defined for a.google.com and b.google.com to be duplicated in here - see line 192. as such, virtual hosts domains need to be pretty specific 

if the host had two route hostnames *.google.com and example.com - this would not be an appropriate domain for the virtual host, as a.google.com should not go in this virtual host.

Reply:
there is currently a order defined in envoy
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-virtualhost
virtual host with domains with exact match such as `a.example.com` will be selected over another virtual domain with a domain set to `*.example.com`

These domains should hopefully be computed correctly in the gateway api layer.
Lets say we have 1 gateway with hostnames set to - `*.example.com`
And 3 HTTPRoutes set to it each with hostnames set to `nil`, `a.example.com`, `b.example.com` respectively
For this case the gateway api should populate the IRRoute.Hostnames to `*.example.com`, `a.example.com` & `b.example.com` for each IR Route respectively.
When these map to xds VirtualHost, Envoy will perform the best match/longest match.

So I guess this boils down to, is https://github.com/envoyproxy/gateway/blob/8b52159c21b1c64ee42ed222bba6e91a343019ab/internal/gatewayapi/helpers.go#L314 doing the right thing ?

Reply:
let me try and give a better example of the case i am concerned about. maybe i am getting myself mixed up, but let me know what you think

lets say i have a route `first-route` with domain `*` and no other matches
and another route `second-route` with domain `foo.example.com` which matches path prefix `/bar`

under the domain order rules, `second-route` takes precedence over `first-route` as it has a non wildcard domain. however, that doesn't mean that `first-route` can never match on host `foo.example.com`! if i request `foo.example.com/baz` then i should not match against `second-route`, and i need to instead match `first-route`. The way this is achieved in the current implementation with one virtual host is that there is a virtual host for the listener ie `*` containing `second-route` and `first-route`, and then `second-route` has an authority check so that even though it is ordered before `first-route`, it wont match unintended domains

in a world with multiple virtual hosts, envoy will match against the `foo.example.com` virtual host, because, as you mentioned, it is more specific. in the `/baz` request case, it will not match against `second-route`, and if `first-route` is not present, it will 404. we cannot fall through between virtual hosts, to my understanding - this is the critical difference from the current approach. once a virtual host is selected, all possible routes need to be present in that virtual host. this is why we need to duplicate `first-route` into the `foo.example.com` virtual host and the `*` virtual host (which should only contain `first-route` in this case)

Reply:
Ah, I see the issue, your statement
```
under the domain order rules, second-route takes precedence over first-route as it has a non wildcard domain. however, that doesn't mean that first-route can never match on host foo.example.com! if i request foo.example.com/baz then i should not match against second-route, and i need to instead match first-route. 
```

does match the intended precedence based outcome defined in the spec, 
https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRoute
rather than a `hostname` ownership approach
```
In the event that multiple HTTPRoutes specify intersecting hostnames (e.g. overlapping wildcard matching and exact matching hostnames), precedence must be given to rules from the HTTPRoute with the largest number of:

Characters in a matching non-wildcard hostname.
Characters in a matching hostname.
If ties exist across multiple Routes, the matching precedence rules for HTTPRouteMatches takes over.
```

a. have we crosschecked if xds virtualHost cannot fall back to a less specific domain match, after a missing route match ?

this logic inside the xds layer seems complex, and will be hard to maintain imo, we've managed to deal with all the complexity in the gateway-api and IR layers and keep xds as simple as possible.
is it possible to move the route duplication logic into gateway api layer if a. is not possible ?

Reply:
@zirain @howardjohn im guessing istio is going to have this same problem

Reply:
Hmm never considered it...

Reply:
Actually re-reading it I think Istio/envoy handles it correctly. You are not supposed to "fall back to a less specific domain match, after a missing route match ?", and Envoy will not

Reply:
@howardjohn from https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRoute the Gateway API spec does define HTTPRoute hostname matching in terms of `precedence`, not `exclusivity`

Reply:
To me its ambiguous. I read it as once you match you should not backtrack, but I can see how you could intepret it otherwise.

This is what leads me to think that:

> If ties exist across multiple Routes, the matching precedence rules for HTTPRouteMatches takes over.

Which implies you only do this on ties, to me. Also I think most LBs work this way so I made that assumptions.

Lets clear it up in the spec and conformance?

Reply:
sure, raising an upstream issue, thanks for cross checking

Reply:
raised https://github.com/kubernetes-sigs/gateway-api/discussions/2294

Reply:
oh, this is interesting! i agree that things would be a lot easier if we had exclusivity on domains like envoy essentially intends. this precedence things is a bit of a pain - but its the current implementation's behaviour so I felt best to keep it

Reply:
ok fab! ill update this

Reply:
thinking a bit more @arkodg , i think as long as there can be a 'tie' between different httproutes that eg both match on google.com or both match on *, then we need to:
1. merge routes with the same domain into one virtual host, with the routes in the correct order within that VH to reflect the httproutematch ordering rules - i think we can probably do this as we convert into IR, so a bit earlier than in this PR
2. a virtual host per domain, not just per route (with multiple domains) because the above merging needs to be done per conflicting domain